stream-tests/src/test/scala/org/apache/pekko/stream/scaladsl/FlowMapWithResourceSpec.scala

```diff
@@ -410,6 +410,24 @@ class FlowMapWithResourceSpec extends StreamSpec(UnboundedMailboxConfig) {
       Await.result(promise.future, 3.seconds) shouldBe Done
     }
 
+    "will close the autocloseable resource" in {
+      val closedCounter = new AtomicInteger(0)
+      val create = () => new AutoCloseable {
+        override def close(): Unit = closedCounter.incrementAndGet()
+      }
+      val (pub, sub) = TestSource
+        .probe[Int]
+        .mapWithResource(create, (_, count) => count)
+        .toMat(TestSink.probe)(Keep.both)
+        .run()
+      sub.expectSubscription().request(2)
+      pub.sendNext(1)
+      sub.expectNext(1)
+      pub.sendComplete()
+      sub.expectComplete()
+      closedCounter.get shouldBe 1
```

Comment:
is there any chance that this test could check the function runs before the close? I would create a class that is AutoCloseable - the close counts the close calls, but the function that the class exposes is set up to check the close count and if it is non-zero, the function should fail.

Reply:
and we need tests where the function fails so that we can see that the close still happens (but only after the function fails)

Reply:
✅ updated, I enhance test to check AutoCloseable is closed after stream is completed~!

Also I add more test about upstream/downstream failure.